[V1][Kernel] Add triton implementation for `reshape_and_cache_flash`

bringlein opened the pull request:
## Purpose

This PR adds a triton implementation of the `reshape_and_cache` kernel, this helps reducing the dependency on non-pytorch/triton kernels for the `triton_attn` kernel. 

The kernel itself has the same (or slightly better) performance on H100 and MI300 as the CUDA kernel. 
<img width="2013" height="798" alt="image" src="https://github.com/user-attachments/assets/cc0d1b1f-1a09-46b3-90e0-f7895baf5df5" />
<img width="2013" height="798" alt="image" src="https://github.com/user-attachments/assets/3f4dabee-76c2-409a-a9d3-2d8bc71c7acc" />


CC: @tdoublep @SageMoore @jvlunteren @jikunshang @cyang49 

## Test Plan

unit tests and end-to-end correctness tests

## Test Result

unit tests: 
```
python3 -m pytest tests/kernels/attention/test_cache.py::test_reshape_and_cache_flash
....
216 passed, 72 skipped in 79.67s (0:01:19) 
```

end2end correctness tests:
```
VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500
```

on main with H100:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.798|±  |0.0180|
|     |       |strict-match    |     5|exact_match|↑  |0.782|±  |0.0185|

```

with this PR using H100:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.800|±  |0.0179|
|     |       |strict-match    |     5|exact_match|↑  |0.784|±  |0.0184|
```

(results are the same if running on MI300). 

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

bringlein committed:
adding reshape and cache kernel, adapting and checking unit tests

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
cosmetics

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
integrate triton reshape and cache into triton_attn

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Merge remote-tracking branch 'origin' into ngl_triton_reshape_and_cache_pr

bringlein committed:
ruff...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fix unittest

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
ruff...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Merge remote-tracking branch 'origin/main' into ngl_triton_reshape_and_cache_pr

bringlein committed:
fixing stupidity

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

cyang49 committed:
ensuring co-authors

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
disable triton kernel if using fp8 on rocm

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fmt...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

yewentao256 reviewed (commented):
Thanks for the work!
Could you also test on the E2E throughput benchmark? `vllm bench throughput ...`
And I don't fully understand why we don't want to use the `reshape_and_cache_flash` CUDA kernel, could you show more context here?

bringlein committed:
simplifying fp8 logic, supporting now float8_e4m3fnuz

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fmt...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Merge branch 'main' into ngl_triton_reshape_and_cache_pr

yewentao256 committed:
Merge branch 'main' into ngl_triton_reshape_and_cache_pr

SageMoore reviewed (approved):
Looks reasonable @bringlein

yewentao256 reviewed (changes requested):
Thanks for the work!

I have tested locally using `vllm-source/benchmarks/kernels/benchmark_reshape_and_cache_flash.py`

Here is what I get on B200:

```bash
  num_tokens  layout      latency_cuda (µs)    latency_triton (µs)
------------  --------  -------------------  ---------------------
           2  NHD                     6.902                 56.777
           4  NHD                     6.714                 55.243
           8  NHD                     6.519                 55.221
          16  NHD                     6.273                 54.965
          32  NHD                     6.355                 55.046
          64  NHD                     6.399                 56.348
         128  NHD                     6.546                 55.745
         256  NHD                     6.431                 54.853
         512  NHD                    10.428                 54.352
        1024  NHD                    25.994                 55.091
        2048  NHD                    47.105                 55.002
        4096  NHD                    87.363                101.223
        8192  NHD                   168.325                211.682
       16384  NHD                   329.958                424.305
       32768  NHD                   654.033                843.652
       65536  NHD                  1303.03                1690.64
```

No HND results since the triton kernel doesn't support it.

From both speed view or from functional view, I don't think it is a good idea to replace the whole Cuda kernel.

bringlein commented:
Hi @yewentao256, 

thanks for your feedback and engagement! 

Regarding your performance numbers: I guess you have measured it without cuda graphs? In this case, you measure the software overhead of the JIT compiler of Triton and that's why it has more or less the very same performance for 2 and 2048 tokens. If you measure it with cuda graphs, i.e. ensuring to not measure the software overhead, I expect that you see very similar performance between cuda and triton kernel (as it is shown in the plots in the initial post). 
(what wonders me is why you see different performance for 65k tokens, there the overhead should not be visible, even without cuda graphs...but the triton performance on blackwell is known to not be very good yet). 

regarding the layout: The HND layout can be implemented, but I'm unsure if it is necessary? 

In my current understanding, we don't want to replace the cuda kernel. We only want to add a triton kernel for the triton backend to have a full triton path. And in the triton backend, we don't have the `HND` layout, or am I missing smth?

bringlein committed:
Merge remote-tracking branch 'origin/main' into ngl_triton_reshape_and_cache_pr and fix merge conflict

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
fmt...

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
update reshape and cache benchmark

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

bringlein committed:
Merge branch 'ngl_triton_reshape_and_cache_pr' of github.com:bringlein/vllm into ngl_triton_reshape_and_cache_pr

bringlein commented:
@mgoin @yewentao256 I updated `benchmarks/kernels/benchmark_reshape_and_cache_flash.py` and added two new flags `implementation` (i.e. cuda or triton) and `mode` (i.e. cuda graphs or not). 

On an H100, I get the following results:
```
python3 ./benchmarks/kernels/benchmark_reshape_and_cache_flash.py --implementation=cuda
Benchmark results for implementation cuda (measuring with cudagraph):
  num_tokens  layout      latency (µs)
------------  --------  --------------
           2  NHD               12.393
           4  NHD               12.789
           8  NHD               14.348
          16  NHD               26.702
          32  NHD               12.422
          64  NHD               12.187
         128  NHD               12.93
         256  NHD               16.437
         512  NHD               37.987
        1024  NHD               59.177
        2048  NHD              104.466
        4096  NHD              195.447
        8192  NHD              378.028
       16384  NHD              741.654
       32768  NHD             1469.05
       65536  NHD             2925.6
```

and for triton
```
python3 ./benchmarks/kernels/benchmark_reshape_and_cache_flash.py --implementation=triton
Benchmark results for implementation triton (measuring with cudagraph):
  num_tokens  layout      latency (µs)
------------  --------  --------------
           2  NHD                8.763
           4  NHD                9.735
           8  NHD               16.91 
          16  NHD               21.476
          32  NHD                9.13
          64  NHD               10.125
         128  NHD               11.34 
         256  NHD               16.787
         512  NHD               35.567
        1024  NHD               56.811
        2048  NHD              103.081
        4096  NHD              199.439
        8192  NHD              395.063
       16384  NHD              766.764
       32768  NHD             1526.15 
       65536  NHD             3050.05 
```
...which is what I would expect, and also is shown in the plots above.

mgoin commented:
Please fix the precommit first
```
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:89:13: F821 Undefined name `key`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:90:13: F821 Undefined name `value`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:91:13: F821 Undefined name `key_cache`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:92:13: F821 Undefined name `value_cache`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:93:13: F821 Undefined name `slot_mapping`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:100:13: F821 Undefined name `key`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:101:13: F821 Undefined name `value`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:102:13: F821 Undefined name `key_cache`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:103:13: F821 Undefined name `value_cache`
Error: benchmarks/kernels/benchmark_reshape_and_cache_flash.py:104:13: F821 Undefined name `slot_mapping`
```

bringlein committed:
mute ruff false-positive

Signed-off-by: Burkhard Ringlein <ngl@zurich.ibm.com>

mgoin reviewed (approved):
LGTM, thanks!

yewentao256 reviewed (approved):
Thanks for the work!
